### ISSUE TITLE
Designing Maps on HTTP Server

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#7263]
I'd like to kick off the [Maps-On-HTTP-Server project](https://github.com/triplea-game/triplea/projects/17
) with this initial design issue.

I'm thinking we'll want a totally separate HTTP server to handle all things maps related. This is really just for clean code separation and possibly load balancing considerations. If the server is doing for example a lot of uploads/downloads, and is handling PNG optimization, then we may need to do some interesting things to scale network, disk and CPU usage. 

With that in mind, I think an easy starting point is simply serving the list of available maps. In essence we'll set up the server, add a new DB schema for maps, load the DB with data and then create an HTTP endpoint to fetch the list of maps.

Simplistically, that looks like this:

![maps](https://user-images.githubusercontent.com/12397753/88500745-68d77100-cf7e-11ea-811d-6271da4df7ca.png)

For fetching and reading maps information, we create a job to do that and set up our data source to be the current YAML file:
![add_maps](https://user-images.githubusercontent.com/12397753/88500764-7b51aa80-cf7e-11ea-9f88-996d0b28aadd.png)

**Comment by RoiEXLab** [comment]
> I'm thinking we'll want a totally separate HTTP server to handle all things maps related.

Agreed. I think it has little to do with lobby and having it separate would be cleaner IMO.
However if we want to use the same auth mechanism like the lobby at some point I really would consider a 'microservice architecture' like the one I mentioned on the forum. I don't think having 2 servers re-use the same code to connect to the same DB would be a good move, mainly because updating the code would only allow for synchronous updating of both services.

> I think an easy starting point is simply serving the list of available maps.

I would actually go one step further with this. In the end the 'StuKa' server should also serve and store the maps, right?
So I would actually change the system, so it still has a list of all maps, but also provides a "download" endpoint that simply acts as a proxy for GitHub, or alternatively simply redirects to the current GitHub link.
These approaches have the benefit that how the server handles map storage becomes an implementation detail, but the interface doesn't need to change for the forseeable future, which would introduce fewer compatibility issues.

> we create a job to do that and set up our data source to be the current YAML file

If you're talking about a cronjob, I'd actually suggest a push system, that is triggered (using GitHub actions for example), whenever the yaml file changes. This way we could have a fluent transition from switching from a yaml based source to something else (perhaps repository based?), without changing the implementation on the MapsServer.

WDYT?

**Comment by DanVanAtta** [comment]
Re: microservices

- The architecture has suffered from a lot of hype. "Amazon has done this, they are really successful, ergo, we'll be successful if we do that". I've read enough blog posts from big companies about how they have fixed their architecture by reverting back to monoliths.

- microservices solve team dependency and deployment synchronization problems while introducing other problems. The problems microservices solve, I don't think we really have them.

- to have microservices you need:
  - good ops, good montoring

Currently we're barely able to even look at our logs, let alone do effective ray-tracing. We have no idea how busy each machine is, etc.

- where do we start drawing the line around services? How long before we get to a state where the tiniest functions are a lambda? Overall I'm thinking we'll have just a few large services, partitioning only when different systems just have nothing to do with each other. For example, if we do a variant of ELO, would that become another service? IMO not, just tie it into the lobby. For now I'm extremely happy letting the lobby also handle user identity.

> . I don't think having 2 servers re-use the same code to connect to the same DB would be a good move

I was thinking Stuka would do an API call, the lobby server does have a login endpoint available.


## Re: map storage

> I would actually go one step further with this. In the end the 'StuKa' server should also serve and store the maps, right?

I was thinking it would only vend the URL. We do not want all network traffic going through Stuka. The location of the URL can then become quite arbitrary, we could make use of the huge amount of disk space we have on the bot servers and use sections of the bot server hard disks to store maps. Even if the URL is not just github, we'd be partitioning the network traffic to storage servers.

> These approaches have the benefit that how the server handles map storage becomes an implementation detail,

I think we would have that. The way the server determines the download URL becomes a detail and potentially dynamic. The database notably tracks that URL, and that could then change.


## re: Sync jobs

> f you're talking about a cronjob,

It's pretty easy to set up a timed action in DropWizard, I was thinking this. Overall that sync action is a shim. Github actions I don't think woudl necessarily be simple. I'm concerned of the map update process on website though and how (or even if) we maintain that.


>  This way we could have a fluent transition from switching from a yaml based source to something else (perhaps repository based?)

I don't know if I see that. The map uploads would all be completely in-game, there would be no file interaction driven by a user. The map upload would be completely driver by API calls.

**Comment by RoiEXLab** [comment]
> I was thinking Stuka would do an API call, the lobby server does have a login endpoint available.

Makes sense, I'm concerned that this will make it harder to push updates to our systems though

> I was thinking it would only vend the URL. 

Ok, in this case I would actually vote for a HTTP redirect. Would "save" us from the traffic because it essentially just serves a URL but leave room for a lot of other choices.

> I am concerned about how the website map update process would continue to work

I think that's a relatively minor thing actually. The data will still be there after all. We might no longer be able to use a system that updates the website "on-change", but setting up a cronjob (using travis or github actions, doesn't really matter) to fetch the up-to-date data every day shouldn't be a problem at all.

**Comment by DanVanAtta** [comment]
What's the additional benefit to a redirect compared to returning the download URL directly?

Today that is essentially what we do, a game client reads all of `triplea_maps.yaml` and then it knows the URL for each map.  If we do a redirect, we'd have to build the endpoint for it, seems extra.

The URLs being served are generated by the server, the source becomes opaque and potentially dynamic, I'm curious what additional flexibility you would see if we did a redirect instead.

------------


> We might no longer be able to use a system that updates the website "on-change", but setting up a cronjob (using travis or github actions, doesn't really matter) to fetch the up-to-date data every day shouldn't be a problem at all.

**Comment by DanVanAtta** [comment]
> We might no longer be able to use a system that updates the website "on-change", but setting up a cronjob (using travis or github actions, doesn't really matter) to fetch the up-to-date data every day shouldn't be a problem at all.

Good point, we'd need to maintain the APIs to serve that list; that could be a hindrance. It would be great to see how much actual usage the maps page gets on the website (is it worht it?)

**Comment by RoiEXLab** [comment]
> What's the additional benefit to a redirect compared to returning the download URL directly?

Less code on the client side.

Instead of having to write code that makes a request to `/get_download_url` and the to `/actual_url` we could simply write code that makes a request to a single endpoint `/download/map123`the automatically redirects to `/actual_url`.
If at some point we decide to serve maps directly via this endpoint because maps are now stored in the DB or on the file system we could silently drop the redirect and the traffic would go over this server

**Comment by DanVanAtta** [comment]
I was not planning/thinking for there to be a `/get_download_url`, instead just a `/get_maps` URL that returns a payload with all maps and their download addresses. A client would then know where to download any given map. The intent is also to help easily replace the existing, instead of downloading `triplea_maps.yaml` and parsing the YAML, we'd just return an equivalent JSON.
